### PR TITLE
docs(virtual-scroll): add initial documentation

### LINKED ISSUE #10123: virtual-scroll: add docs
Add documentation for virtual scrolling

### PR DESCRIPTION
**Pull request by mmalerba**
fixes #10123

### CODE REVIEW

File: src/cdk-experimental/scrolling/virtual-scroll.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+`<cdk-virtual-scroll-viewport>` is used to display a scrolling list containing a large number of
+items that would have a negative effect on performance if all rendered at once. Instead the virtual
+scroll viewport renders each item as needed when the user scrolls it into view. The
+`<cdk-virtual-scroll-viewport>` works together with the `*cdkVirtualFor` directive which is used to
+render items inside of the viewport. It also requires that a `VirtualScrollStrategy` be provided.
+The easiest way to provide a strategy is with one of the directives `itemSize` or `autosize`.
+
+For some example usages, 
+[see the demo app](https://github.com/angular/material2/tree/master/src/demo-app/virtual-scroll).
```
Comment (crisbeto) [inline src/cdk-experimental/scrolling/virtual-scroll.md]: Add a TODO here to add live examples once this lands in stable?

File: src/cdk-experimental/scrolling/virtual-scroll.md [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+### Creating items in the viewport
+`*cdkVirtualFor` should be used instead of `*ngFor` to create items in the
+`<cdk-virtual-scroll-viewport>`. `*cdkVirtualFor` supports most of the same features as `*ngFor`.
```
Comment (crisbeto) [inline src/cdk-experimental/scrolling/virtual-scroll.md]: Should it mention which features it doesn't support?

File: src/cdk-experimental/scrolling/virtual-scroll.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+</cdk-virtual-scroll-viewport>
+```
+
+A `trackBy` function can be specified and works the same as the `*ngFor` `trackBy`. The `index`
+passed to the tracking function will be the index in the data source, not the index in the rendered
+portion.
+
+#### View recycling
+In order to improve performance `*cdkVirtualFor` saves the views it creates in a cache when they are
+no longer needed. This way the next time a new view is needed once can be recycled from the cache
```
Comment (crisbeto) [inline src/cdk-experimental/scrolling/virtual-scroll.md]: The "once" seems out of place here.

File: src/cdk-experimental/scrolling/virtual-scroll.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+rather than being created from scratch. The size of this cache can be adjusted using the
+`templateCacheSize` input. The cache size defaults to `20` and caching can be disabled by setting it
```
Comment (crisbeto) [inline src/cdk-experimental/scrolling/virtual-scroll.md]: IMO we should avoid specifying defaults in the docs since they could change and we'll probably forget to update them.

File: src/cdk-experimental/scrolling/virtual-scroll.md [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+`*cdkVirtualFor` is set up to accept data from an `Array`, `Observable`, or `DataSource`. The
+`DataSource` for the virtual scroll is the same one used by the table and tree components. A
+`DataSource` is simply an abstract class that has two methods: `connect` and `disconnect`. The
+`connect` method will be called by the virtual scroll viewport to receive a stream that emits the
+data array that should be rendered. The viewport will call `disconnect` when the viewport is
+destroyed, which may be the right time to clean up any subscriptions that were registered during the
+connect process.
+
+### Scrolling over fixed size items
+If you're scrolling over a list of items that are all the same fixed size, you can use the
```
Comment (crisbeto) [inline src/cdk-experimental/scrolling/virtual-scroll.md]: Mention the advantages/disadvantages of using a fixed-size viewport?

File: src/cdk-experimental/scrolling/virtual-scroll.md (hunk 2 above)
Comment (mmalerba) [reply src/cdk-experimental/scrolling/virtual-scroll.md]: I wrote that because I'm naturally hedgey, but I think it supports all of the same features as `*ngFor`, updated to say that.

File: src/cdk-experimental/scrolling/virtual-scroll.md (hunk 3 above)
Comment (mmalerba) [reply src/cdk-experimental/scrolling/virtual-scroll.md]: whoops, was supposed to read "one"

File: src/cdk-experimental/scrolling/virtual-scroll.md (hunk 4 above)
Comment (mmalerba) [reply src/cdk-experimental/scrolling/virtual-scroll.md]: I think its useful for people to know. I'll put it in the jsdoc rather than here though.

File: src/cdk-experimental/scrolling/virtual-scroll.md [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+  </div>
+</cdk-virtual-scroll-viewport>
+```
+
+A `trackBy` function can be specified and works the same as the `*ngFor` `trackBy`. The `index`
+passed to the tracking function will be the index in the data source, not the index in the rendered
+portion.
+
+#### View recycling
+In order to improve performance `*cdkVirtualFor` saves the views it creates in a cache when they are
```
Comment (jelbourn) [inline src/cdk-experimental/scrolling/virtual-scroll.md]: Editing for brevity and passive voice:
```markdown
To improve rendering performance, `*cdkVirtualFor` caches previously created views after
they are no longer needed. When a new view would normally be created, a cached view
is reused instead. The size of the view cache can be adjusted via the `templateCacheSize`
property; setting this size to `0` disables caching.
```

It would also be good to explain _why_ you might want to change the cache size.

File: src/cdk-experimental/scrolling/virtual-scroll.md [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+rather than being created from scratch. The size of this cache can be adjusted using the
+`templateCacheSize` input. Caching can be disabled by setting it to `0`.
+
+```html
+<cdk-virtual-scroll-viewport itemSize="50">
+  <div *cdkVirtualFor="let item of items; templateCacheSize: 0">{{item}}</div>
+</cdk-virtual-scroll-viewport>
+```
+
+#### Specifying items with an Observable or DataSource
```
Comment (jelbourn) [inline src/cdk-experimental/scrolling/virtual-scroll.md]: Just `#### Providing data`?

File: src/cdk-experimental/scrolling/virtual-scroll.md [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+`*cdkVirtualFor` is set up to accept data from an `Array`, `Observable`, or `DataSource`. The
```
Comment (jelbourn) [inline src/cdk-experimental/scrolling/virtual-scroll.md]: * `is set up to accept` -> `accepts`
* The `Observable` should probably be `Obervable<Array>`

Not for now, but now that this uses a `DataSource` as well, we should really have a separate page for that.

File: src/cdk-experimental/scrolling/virtual-scroll.md [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+`DataSource` for the virtual scroll is the same one used by the table and tree components. A
+`DataSource` is simply an abstract class that has two methods: `connect` and `disconnect`. The
+`connect` method will be called by the virtual scroll viewport to receive a stream that emits the
+data array that should be rendered. The viewport will call `disconnect` when the viewport is
+destroyed, which may be the right time to clean up any subscriptions that were registered during the
+connect process.
+
+### Scrolling over fixed size items
+If you're scrolling over a list of items that are all the same fixed size, you can use the
```
Comment (jelbourn) [inline src/cdk-experimental/scrolling/virtual-scroll.md]: I would rephrase `When all items are the same fixed size, ...`

(similar for autosize section)

File: src/cdk-experimental/scrolling/virtual-scroll.md [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+`AutoSizeVirtualScrollStrategy`. This can be added to your viewport by using the `autosize`
+directive.
+
+```html
+<cdk-virtual-scroll-viewport autosize>
+  ...
+</cdk-virtual-scroll-viewport>
+```
+
+The `autosize` strategy allows the buffer to be configured through two inputs `minBufferPx` and
```
Comment (jelbourn) [inline src/cdk-experimental/scrolling/virtual-scroll.md]: I would break up the [paragraph structure a bit here:
```markdown
The `autosize` strategy is configured through two inputs: `minBufferPx` and `addBufferPx`.

**`minBufferPx`** determines the minimum space outside virtual scrolling viewport that will be filled
with content. Increasing this will increase the amount of content a user will see before more content
must be rendered. However, too large a value will cause more content to be rendered than is necessary.

**`addBufferPx`** determines the amount of content that will be added incrementally as
the viewport is scrolled. This should be greater than the size of `minBufferPx` so that one
"render" is needed at a time.
```

File: src/cdk-experimental/scrolling/virtual-scroll.md [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+```html
+<cdk-virtual-scroll-viewport itemSize="50" bufferSize="1">
+  ...
+</cdk-virtual-scroll-viewport>
+```
+
+**Note: The fixed size strategy will likely be changed to allow specifying a separate
+`minBufferPx` and `addBufferPx` like the autosize strategy**
+
+The fixed size viewport strategy is less flexible than the autosize strategy because it requires all
```
Comment (jelbourn) [inline src/cdk-experimental/scrolling/virtual-scroll.md]: I would just explain this once near the beginning of the doc; it's currently mentioned a few times throughout

File: src/cdk-experimental/scrolling/virtual-scroll.md [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+```html
+<cdk-virtual-scroll-viewport autosize minBufferPx="50" addBufferPx="100">
+  ...
+</cdk-virtual-scroll-viewport>
+```
+
+Because the auto size strategy needs to measure the size of the elements, its performance may not
+be as good as the fixed size strategy. 
+
+### Setting the viewport orientation
```
Comment (jelbourn) [inline src/cdk-experimental/scrolling/virtual-scroll.md]: Just `#### Viewport orientation` ?

File: src/cdk-experimental/scrolling/virtual-scroll.md [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+The orientation of the viewport can be adjusted by using the `orientation` input. It defaults to
+`vertical` which virtualizes scrolling along the y-axis. It can be set to `horizontal` to virtualize
+scrolling along the x-axis. If you use this option you need to make sure that the content is
+actually laid out so that it extends along the x-axis. To do this you may want to target CSS at
+`.cdk-virtual-scroll-content-wrapper` which is the wrapper element that contains the rendered
```
Comment (jelbourn) [inline src/cdk-experimental/scrolling/virtual-scroll.md]: Would it be better/possible to recommend styling their own element instead of a class built into the virtual scroller?

File: src/cdk-experimental/scrolling/virtual-scroll.md [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+<cdk-virtual-scroll-viewport autosize minBufferPx="50" addBufferPx="100">
+  ...
+</cdk-virtual-scroll-viewport>
+```
+
+Because the auto size strategy needs to measure the size of the elements, its performance may not
+be as good as the fixed size strategy. 
+
+### Setting the viewport orientation
+The orientation of the viewport can be adjusted by using the `orientation` input. It defaults to
```
Comment (jelbourn) [inline src/cdk-experimental/scrolling/virtual-scroll.md]:
```markdown
The virtual-scroll viewport defaults to a vertical orientation, but can also be set to
`orientation="horizontal"`. When changing the orientation, ensure that the item are laid
out horizontally via CSS.
```

File: src/cdk-experimental/scrolling/virtual-scroll.md [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+
+`<cdk-virtual-scroll-viewport>` is used to display a scrolling list containing a large number of
+items that would have a negative effect on performance if all rendered at once. Instead the virtual
+scroll viewport renders each item as needed when the user scrolls it into view. The
+`<cdk-virtual-scroll-viewport>` works together with the `*cdkVirtualFor` directive which is used to
+render items inside of the viewport. It also requires that a `VirtualScrollStrategy` be provided.
+The easiest way to provide a strategy is with one of the directives `itemSize` or `autosize`.
+
+<!-- TODO(mmalerba): Add embedded examples once this is available in @angular/cdk -->
+For some example usages,
```
[the remark this answers is not in the document]
Comment (Zizzamia) [reply src/cdk-experimental/scrolling/virtual-scroll.md]: At some point, it would be great have a link to a live **stackblitz** implementation.

File: src/cdk-experimental/scrolling/virtual-scroll.md (hunk 13 above)
Comment (mmalerba) [reply src/cdk-experimental/scrolling/virtual-scroll.md]: It's eventually going to be possible to specify that element themselves, or just allow it to be implicitly created if they don't specify. That's part of the work for making it work with `<table>`, `<ul>`, etc.

File: src/cdk-experimental/scrolling/virtual-scroll.md [hunk 16]
```diff
@@ -0,0 +1,141 @@
+**Warning: this component is still experimental. It may have bugs and the API may change at any
+time**
+
+`<cdk-virtual-scroll-viewport>` displays large lists of elements performantly by only
+rendering the items that appear on-screen. This components works with `*cdkVirtualFor`,
```
Comment (jelbourn) [inline src/cdk-experimental/scrolling/virtual-scroll.md]: `appear` -> `fit` (just occurred to me on second reading)

File: src/cdk-experimental/scrolling/virtual-scroll.md [hunk 17]
```diff
@@ -0,0 +1,139 @@
+**Warning: this component is still experimental. It may have bugs and the API may change at any
+time**
+
+`<cdk-virtual-scroll-viewport>` displays large lists of elements performantly by only
```
Comment (amcdnl) [inline src/cdk-experimental/scrolling/virtual-scroll.md]: I think we need to explain virtual scroll a little bit. I don't think people understand the difference between virtual scroll and infinite scroll. A idea gist.

> Virtual scrolling enables developers to render large lists of data in a performant way. Loading hundreds of elements can be slow in any browser, virtual scrolling enables a performant want to simulate all items are rendered by making the height of the container element the total number of elements to be rendered and then only rendering the items in view. Virtual scroll is different from strategies like infinite scroll where it renders a set amount of elements and then when you hit the bottom renders the rest.

File: src/cdk-experimental/scrolling/virtual-scroll.md [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+rendering the items that fit on-screen. This components works with `*cdkVirtualFor`,
+a virtualized replacement for `*ngFor`.
+
+<!-- TODO(mmalerba): Add embedded examples once this is available in @angular/cdk -->
+For some example usages, 
+[see the demo app](https://github.com/angular/material2/tree/master/src/demo-app/virtual-scroll).
+
+### Creating items in the viewport
+`*cdkVirtualFor` replaces `*ngFor` inside of a `<cdk-virtual-scroll-viewport>`, supporting the exact
```
Comment (amcdnl) [inline src/cdk-experimental/scrolling/virtual-scroll.md]: The replacement is mentioned right above this. Perhaps remove from one or the other?

File: src/cdk-experimental/scrolling/virtual-scroll.md [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+`*cdkVirtualFor` accepts data from an `Array`, `Observable<Array>`, or `DataSource`. The
+`DataSource` for the virtual scroll is the same one used by the table and tree components. A
+`DataSource` is simply an abstract class that has two methods: `connect` and `disconnect`. The
+`connect` method will be called by the virtual scroll viewport to receive a stream that emits the
+data array that should be rendered. The viewport will call `disconnect` when the viewport is
+destroyed, which may be the right time to clean up any subscriptions that were registered during the
+connect process.
+
+### Scrolling over fixed size items
+When all items are the same fixed size,, you can use the `FixedSizeVirtualScrollStrategy`. This can
```
Comment (amcdnl) [inline src/cdk-experimental/scrolling/virtual-scroll.md]: Two commas: `fixed size,,`

File: src/cdk-experimental/scrolling/virtual-scroll.md [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+* `last` - Whether this is the last item in the data source.
+* `even` - Whether the `index` is even.
+* `odd` - Whether the `index` is odd.
+
+All of these apply to the index of the item in the data source, not the index in the rendered
+portion of the data.
+
+```html
+<cdk-virtual-scroll-viewport itemSize="50">
+  <div *cdkVirtualFor="let item of items;
```
[the remark this answers is not in the document]
Comment (amcdnl) [reply src/cdk-experimental/scrolling/virtual-scroll.md]: You'd be surprised ....

File: src/cdk-experimental/scrolling/virtual-scroll.md [hunk 21]
```diff
@@ -0,0 +1,139 @@
...
+                       let index = index;
+                       let count = count">
+    {{item}} ({{index}} of {{count}})
+  </div>
+</cdk-virtual-scroll-viewport>
+```
+
+A `trackBy` function can be specified and works the same as the `*ngFor` `trackBy`. The `index`
```
Comment (amcdnl) [inline src/cdk-experimental/scrolling/virtual-scroll.md]: I feel like this `trackBy` is just randomly mentioned at the end here.

File: src/cdk-experimental/scrolling/virtual-scroll.md [hunk 22]
```diff
@@ -0,0 +1,139 @@
...
+passed to the tracking function will be the index in the data source, not the index in the rendered
+portion.
+
+#### View recycling
+To improve rendering performance, `*cdkVirtualFor` caches previously created views after
```
Comment (amcdnl) [inline src/cdk-experimental/scrolling/virtual-scroll.md]: Might be worth mentioning things like portals that are attached to the VCR can cause issues.

File: src/cdk-experimental/scrolling/virtual-scroll.md [hunk 23]
```diff
@@ -0,0 +1,139 @@
...
+
+### Viewport orientation
+The virtual-scroll viewport defaults to a vertical orientation, but can also be set to
+`orientation="horizontal"`. When changing the orientation, ensure that the item are laid
+out horizontally via CSS. To do this you may want to target CSS at
+`.cdk-virtual-scroll-content-wrapper` which is the wrapper element that contains the rendered
+content.
+
+```html
+<cdk-virtual-scroll-viewport autosize orientation="horizontal">
```
Comment (amcdnl) [inline src/cdk-experimental/scrolling/virtual-scroll.md]: I'd generally discourage users from using autosize, thus removing it as the default on examples like this.

File: src/cdk-experimental/scrolling/virtual-scroll.md [hunk 24]
```diff
@@ -0,0 +1,139 @@
...
+advantage of this constraint is that it allows for better performance, since items do not need to be
+measured as they are rendered. 
+
+```html
+<cdk-virtual-scroll-viewport itemSize="50">
+  ...
+</cdk-virtual-scroll-viewport>
+```
+
+The fixed size strategy also supports setting the buffer size, i.e. the number of items rendered
```
Comment (amcdnl) [inline src/cdk-experimental/scrolling/virtual-scroll.md]: Whats the default buffer size?

File: src/cdk-experimental/scrolling/virtual-scroll.md [hunk 25]
```diff
@@ -0,0 +1,139 @@
...
+### Scrolling over fixed size items
+When all items are the same fixed size,, you can use the `FixedSizeVirtualScrollStrategy`. This can
+be easily added to your viewport using the `itemSize` directive. The fixed size viewport strategy is
+less flexible than the autosize strategy because it requires all items to be the same size, but the
+advantage of this constraint is that it allows for better performance, since items do not need to be
+measured as they are rendered. 
+
+```html
+<cdk-virtual-scroll-viewport itemSize="50">
+  ...
```
Comment (amcdnl) [inline src/cdk-experimental/scrolling/virtual-scroll.md]: I think the examples should show the for in them, people like to just copy/paste.

File: src/cdk-experimental/scrolling/virtual-scroll.md [hunk 26]
```diff
@@ -0,0 +1,139 @@
...
+memory you may wish to reduce this number to avoid spending too much memory on the template cache.
+
+```html
+<cdk-virtual-scroll-viewport itemSize="50">
+  <div *cdkVirtualFor="let item of items; templateCacheSize: 0">{{item}}</div>
+</cdk-virtual-scroll-viewport>
+```
+
+#### Specifying data
+`*cdkVirtualFor` accepts data from an `Array`, `Observable<Array>`, or `DataSource`. The
```
Comment (amcdnl) [inline src/cdk-experimental/scrolling/virtual-scroll.md]: I'd show an example code snippet here.

File: src/cdk-experimental/scrolling/virtual-scroll.md (hunk 22 above)
Comment (amcdnl) [reply src/cdk-experimental/scrolling/virtual-scroll.md]: Might even put this at the end too. Its a implementation detail rather than something important for the user to know.

File: src/cdk-experimental/scrolling/virtual-scroll.md (hunk 17 above)
Comment (mmalerba) [reply src/cdk-experimental/scrolling/virtual-scroll.md]: made some slight tweaks and merged with my sentence above

File: src/cdk-experimental/scrolling/virtual-scroll.md (hunk 21 above)
Comment (mmalerba) [reply src/cdk-experimental/scrolling/virtual-scroll.md]: That's because trackBy isn't a context variable, so it didn't fit in the part above, but I do want to mention that it's still supported

File: src/cdk-experimental/scrolling/virtual-scroll.md (hunk 22 above)
Comment (mmalerba) [reply src/cdk-experimental/scrolling/virtual-scroll.md]: It can be important for the user to know depending on their use case, some components may not deal well with having their templates recycled.

It does sound like a good idea to mention somewhere that the ViewContainer is managed by the virtual-scroller and so any components that attempt to do things with it are going to run into issues. I'm not sure where the best place for this is... In the future I also think we could mark the views that are managed by the virtual scroll in order to play better with other components that may modify the ViewContainer

File: src/cdk-experimental/scrolling/virtual-scroll.md (hunk 26 above)
Comment (mmalerba) [reply src/cdk-experimental/scrolling/virtual-scroll.md]: I'll add that once we move it into the cdk proper and have actual embedded examples

File: src/cdk-experimental/scrolling/virtual-scroll.md (hunk 25 above)
Comment (mmalerba) [reply src/cdk-experimental/scrolling/virtual-scroll.md]: I'll add full working embedded example code when I move this out of cdk-experimental

File: src/cdk-experimental/scrolling/virtual-scroll.md (hunk 23 above)
Comment (mmalerba) [reply src/cdk-experimental/scrolling/virtual-scroll.md]: Fair enough, we don't want users putting `autosize` just because its less typing. If they know the `itemSize` it's going to be more performant.

**Comment by mmalerba** [comment]
@jelbourn @amcdnl PTAL

File: src/cdk-experimental/scrolling/virtual-scroll.md (hunk 22 above)
Comment (amcdnl) [reply src/cdk-experimental/scrolling/virtual-scroll.md]: Probably put it at the bottom as a note.
